docs/src/lib/indexing.md

```diff
@@ -76,6 +76,127 @@ For performance reasons, accessing, via `getindex` or `view`, a single `row` and
 * `@view dfr[col]` -> a `0`-dimensional view into `parent(dfr)[DataFrames.row(dfr), col]`;
 * `@view dfr[cols]` -> a `DataFrameRow` with parent `parent(dfr)` if `cols` is a colon and `parent(dfr)[cols]` otherwise;
 
-## `setindex!`
+## `setindex!`, `setproperty!` and broadcasted assignment
+
+This is a planned functionality.
+
+The following list the effect of `setindex!`, `setproperty!` and broadcasted assignment and operations.
 
-Under construction
+In all operations copying is performed in general.
+If it is not performed a description explicitly mentions that the data is assigned *in-place*.
+
+### `DataFrame`:
+
+> `df[col] = v` and `df.col = v`
+
+* `v` must be an `AbstractVector`. If `ncol(df) > 0` then `length(v)` must be equal to `nrow(df)`.
+* `v` is assigned to `df` without copying unless it is a range, in which case it is converted to a `Vector`.
+* `haskey(df, col)` must hold unless `col isa Symbol`, in which case a new column is added at the end of `df`.
+
+> `df[col] .= v` and `df.col .= v`
+
+* `v` must be broadcastable.
+* if `haskey(df, col)` then standard in-place broadcasted assignment of `v` to the column vector `df[col]` is performed.
+* if not `haskey(df, col)` and `col isa Symbol`
+    * `df.col .= v` throws an error.
+    * if `ncol(df) > 0` then a new verctor `c` of an appropriate type is created, having `nrow(df)` entries;
+      then `c .= v` and `df.col = c` is performed.
+    * if `ncol(df) == 0` then same operation as in the case above is performed, whith the following additional rules:
+        * if `v` has 0-dimensions then 1 row is created,
+        * if `v` has 1-dimension the number of rows equal to its length are created,
+        * if `v` has more than 1 dimension then an error is thrown
+* if not `haskey(df, col)` and `!(col isa Symbol)` an error is thrown
+
+> `df[cols] = v`
+
+* if `v` is a `DataFrame` or `NamedTuple` of vectors or `AbstractDict` of vectors:
+    * `nrow(df)` must be equal to number of rows in `v` unless `ncol(df)==0` in
+      which case it is enough that number of rows in each entry of `v` is constant;
+    * number of columns in `v` must be equal to length of `cols`
+    * before making any assignment `cols` is transformed to `Symbols` using `names(df)`
+      (which means that passing `Integer` or `Bool` will fail for nonexistent columns,
+      but adding new columns as `Symbol` is allowed)
+    * then `df[col] = v[col]` is called for each column name `col`
+* if `v` is a vector of vectors or a tuple of vectors the same process is performed but
```

Comment:
I have added this rule. I do not know if we want to keep it (or keep it only for vectors and not for tuples)

Reply:
Sounds fine. Am I right that this doesn't introduce any ambiguity if you want a column which is a vector of vectors (since you would have to wrap it in a vector/tuple/data frame anyway since this is a multi-column assignment)?

Reply:
Right - you would have to wrap it. Single column assignment is another rule.

Actually this is very delicate. My idea is that `df[cols] = v` should ideally give the same result as `df[cols] = DataFrame(v)` just without having to materialize the intermediate `DataFrame`.

The same for `df[rows, cols] = v`.

But now I see that:

```
julia> x = ([1,2,3], [1,2,3])
([1, 2, 3], [1, 2, 3])

julia> DataFrame(x)
ERROR: ArgumentError: 'Tuple{Array{Int64,1},Array{Int64,1}}' iterates 'Array{Int64,1}' values, which don't satisfy the Tables.jl Row-iterator interface

julia> x = [[1,2,3], [1,2,3]]
2-element Array{Array{Int64,1},1}:
 [1, 2, 3]
 [1, 2, 3]

julia> DataFrame(x)
3×2 DataFrame
│ Row │ x1    │ x2    │
│     │ Int64 │ Int64 │
├─────┼───────┼───────┤
│ 1   │ 1     │ 1     │
│ 2   │ 2     │ 2     │
│ 3   │ 3     │ 3     │
```

which is quite surprising (in general the Tables.jl & friends catch all `DataFrame` constructor always catches me off guard 😞). So I think we should drop tuple here and leave only the vector?

Reply:
Why doesn't Tables.jl accept a tuple here?

Reply:
It accepts it, but assumes that the contents is row-wise not col-wise, see e.g.:
```
julia> DataFrame(((a=1,b=2),(a=3,b=4)))
2×2 DataFrame
│ Row │ a     │ b     │
│     │ Int64 │ Int64 │
├─────┼───────┼───────┤
│ 1   │ 1     │ 2     │
│ 2   │ 3     │ 4     │
```
That is why I say that this Tables.jl API is tricky when combined with what we already had in DataFrames.jl.

Reply:
I will remove tuples here from the description.

Reply:
Then maybe it should do the same for vectors?

Reply:
I can remove them too, but actually we have a `DataFrame` constructor taking a vector of vectors, which treats them col-wise.

The general question is in the assignment:

```
df[rows, cols] = v
```
should `v` be treated row-wise or col-wise? And a natural answer is to follow what `DataFrame(v)` would do.

In general I am OK, to narrow down what is accepted now, and we can add the functionality later.

Reply:
I agree. But it doesn't sound very consistent for Tables.jl to treat vectors differently from tuples. Or maybe we should special-case collections of named tuples. Anyway it's probably worth opening a separate issue to discuss that.

Reply:
There is no inconsistency. Tables.jl does not support vector of vectors at all (the same as tuple of vectors). It is only supported in DataFrames.jl specific `DataFrame` constructor.

Actually thinking of it we might want to accept a `Matrix` as `v` (it is also not supported by Tables.jl but is supported by `DataFrame` constructor).

Reply:
OK. Though currently there seems to be a bug with this (but at least it should work since there's a special method for it in DataFrames):
```julia
julia> DataFrame([(a=1,b=2),(a=3,b=4)])
ERROR: UndefVarError: DataValueUnwrapper not defined
```

Regarding what we can do in DataFrames, why not add a constructor taking a tuple of vectors to parallel the one taking a vector of vectors?

> Actually thinking of it we might want to accept a `Matrix` as `v` (it is also not supported by Tables.jl but is supported by `DataFrame` constructor).

Fine with me. But this can be added at any time since that's an error AFAICT.

Reply:
1. Following your suggestion I have commented on the constructors in a separate issue https://github.com/JuliaData/DataFrames.jl/issues/1599.
2. Are you on latest versions of all packages from Tables.jl & friends? For me it works:
```
julia> DataFrame([(a=1,b=2),(a=3,b=4)])
2×2 DataFrame
│ Row │ a     │ b     │
│     │ Int64 │ Int64 │
├─────┼───────┼───────┤
│ 1   │ 1     │ 2     │
│ 2   │ 3     │ 4     │
```
(and even if it did not work it would not be an error in DataFrames.jl 😀)

Reply:
OK, actually I tested that on my PooledArrays branch which was behind master. It works now.

Reply:
Given the discussion in https://github.com/JuliaData/DataFrames.jl/issues/1599#issuecomment-462261917 I changed my mind and would leave only vector of vectors here for now.

The resason is that in general it is not clear if you have a collection if it should be treated row-wise or col-wise. We can add other values of `v` later if there is a need.

Reply:
My suggestion was that the collection type should not matter to decide whether it's row- or column-wise: tuples and vectors should be treated the same. Two constructors would take tuples and vectors of vectors. Other collections would be assumed to be row-oriented and handled by Tables.jl.

Reply:
OK :). So I add tuples here and in the constructors thread.